Assets/MixedRealityToolkit.Providers/ObjectMeshObserver/SpatialObjectMeshObserver.cs

```diff
@@ -0,0 +1,336 @@
+        {
+            if (!sendObservations) { return; }
+
+            IsRunning = true;
+
+            // Get the collection of MeshFilters
+            if (spatialMeshObject != null)
+            {
+                MeshFilter[] meshFilters = spatialMeshObject.GetComponentsInChildren<MeshFilter>();
+                for (int i = 0; i < meshFilters.Length; i++)
+                {
+                    SpatialAwarenessMeshObject meshObject = SpatialAwarenessMeshObject.Create(
+                        meshFilters[i].sharedMesh,
+                        MeshPhysicsLayer,
+                        $"Spatial Object Mesh {currentMeshId}",
+                        currentMeshId,
+                        ObservedObjectParent);
+
+                    meshObject.GameObject.transform.localPosition = meshFilters[i].transform.position;
+                    meshObject.GameObject.transform.localRotation = meshFilters[i].transform.rotation;
+
+                    ApplyMeshMaterial(meshObject);
+
+                    meshes.Add(currentMeshId, meshObject);
+
+                    currentMeshId++;
+                }
+            }
+            sendObservations = false;
+        }
+
+        /// <inheritdoc />
+        public override void Suspend()
+        {
+            IsRunning = false;
+        }
+
+        private void RemoveMeshObject(int id)
+        {
+            SpatialAwarenessMeshObject meshObject = null;
+
+            if (meshes.TryGetValue(id, out meshObject))
+            {
+                // Remove the mesh object from the collection.
+                meshes.Remove(id);
+                if (meshObject != null)
+                {
+                    SpatialAwarenessMeshObject.Cleanup(meshObject);
+                }
+
+                // Send the mesh removed event
+                SpatialAwarenessSystem?.RaiseMeshRemoved(this, id);
+            }
+        }
+
+        #endregion IMixedRealitySpatialAwarenessObserver implementation
+
+        #region IMixedRealitySpatialAwarenessMeshObserver implementation
+
+        public SpatialAwarenessMeshDisplayOptions displayOption = SpatialAwarenessMeshDisplayOptions.Visible;
+        
+        /// <inheritdoc />
+        public SpatialAwarenessMeshDisplayOptions DisplayOption
+        {
+            get => displayOption;
+
+            set
+            {
+                displayOption = value;
+                ApplyUpdatedMeshDisplayOption(displayOption);
+            }
+        }
+
+        /// <inheritdoc />
+        public SpatialAwarenessMeshLevelOfDetail LevelOfDetail { get; set; } = SpatialAwarenessMeshLevelOfDetail.Coarse;
+
+        private Dictionary<int, SpatialAwarenessMeshObject> meshes = new Dictionary<int, SpatialAwarenessMeshObject>();
+
+        /// <inheritdoc />
+        public IReadOnlyDictionary<int, SpatialAwarenessMeshObject> Meshes =>
+            new Dictionary<int, SpatialAwarenessMeshObject>(meshes) as IReadOnlyDictionary<int, SpatialAwarenessMeshObject>;
```

Comment:
This implementation will generate a new dictionary each time the **.Meshes** property is read. Since the contents of this dictionary will be unlikely to change over time, consider caching this instance and returning it for future requests. Of course, if a method like **RemoveMeshObject** is called, the cached dictionary would need to be nullified so that the property accessor would generate it again.

Reply:
that is by design so that the collection does not change during enumeration. same behavior as the WMR Mesh observer

Reply:
for this pr, this code will not change. we should open a discussion re: safety vs performance and guidance relating to the two